File: sdks/local-rp/php/src/Encoding.php
<?php

declare(strict_types=1);

namespace LinkKeys\LocalRp;

use Csilgen\Generated\LocalRpEncryptedCallback;
use Csilgen\Generated\SignedLocalRpLoginRequest;

/**
 * URL-parameter helpers (Wire Precision, "URL and parameter conventions"):
 * every CBOR-in-URL value is base64url-encoded, unpadded, exactly like
 * `crates/liblinkkeys/src/encoding.rs`'s `Base64UrlUnpadded` helpers.
 */
final class Encoding
{
    /** Base64url, no padding. Rejects standard-alphabet input and padded input on decode. */
    public static function base64UrlEncodeUnpadded(string $bytes): string
    {
        return rtrim(strtr(base64_encode($bytes), '+/', '-_'), '=');
    }

    public static function base64UrlDecodeUnpadded(string $s): string
    {
        if (str_contains($s, '=')) {
            throw new \InvalidArgumentException('base64url input must not be padded');
        }
        if (str_contains($s, '+') || str_contains($s, '/')) {
            throw new \InvalidArgumentException('base64url input must use the URL-safe alphabet');
        }
        if (!preg_match('/^[A-Za-z0-9_-]*$/', $s)) {
            throw new \InvalidArgumentException('base64url input contains invalid characters');
        }
        $padded = $s . str_repeat('=', (4 - strlen($s) % 4) % 4);
        $decoded = base64_decode(strtr($padded, '-_', '+/'), true);
        if ($decoded === false) {
            throw new \InvalidArgumentException('invalid base64url input');
        }
        return $decoded;
    }

    /** `GET /auth/local-rp?signed_request=<this>`. */
    public static function signedLocalRpLoginRequestToUrlParam(SignedLocalRpLoginRequest $req): string
    {
        return self::base64UrlEncodeUnpadded(Wire::encodeSignedLocalRpLoginRequest($req));
    }

    public static function signedLocalRpLoginRequestFromUrlParam(string $param): SignedLocalRpLoginRequest
    {
        return Wire::decodeSignedLocalRpLoginRequest(self::base64UrlDecodeUnpadded($param));
    }

    /** The callback redirect's `&encrypted_token=` query-parameter value. */
    public static function localRpEncryptedCallbackToUrlParam(LocalRpEncryptedCallback $callback): string
    {
        return self::base64UrlEncodeUnpadded(Wire::encodeLocalRpEncryptedCallback($callback));
    }

    public static function localRpEncryptedCallbackFromUrlParam(string $param): LocalRpEncryptedCallback
    {
        return Wire::decodeLocalRpEncryptedCallback(self::base64UrlDecodeUnpadded($param));
    }
}

File: sdks/local-rp/dart/lib/src/encoding.dart
// URL parameter encoding helpers -- mirrors
// `crates/liblinkkeys/src/encoding.rs` / the Rust/Go/Java reference SDKs'
// own `encoding` modules. All CBOR-in-URL values are base64url-encoded,
// unpadded (RFC 4648 section 5), matching `base64ct::Base64UrlUnpadded`
// exactly: no standard-alphabet (`+`/`/`) characters, no `=` padding.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'errors.dart';
import 'wire/codec.dart';
import 'wire/types.dart';

String encodeUrlParam(List<int> b) => base64Url.encode(b).replaceAll('=', '');

Uint8List decodeUrlParam(String s) {
  if (s.contains('=')) {
    throw SdkException(SdkExceptionKind.invalidInput,
        'padded base64 is not accepted (expected unpadded base64url)');
  }
  if (s.contains('+') || s.contains('/')) {
    throw SdkException(SdkExceptionKind.invalidInput,
        'standard base64 alphabet is not accepted (expected base64url)');
  }
  try {
    final padded = _addPadding(s);
    return base64Url.decode(padded);
  } on FormatException catch (e) {
    throw SdkException(
        SdkExceptionKind.invalidInput, 'base64url decode failed: ${e.message}',
        cause: e);
  }
}

String _addPadding(String s) {
  final mod = s.length % 4;
  if (mod == 0) return s;
  return s + ('=' * (4 - mod));
}

/// Encode for the begin route's `?signed_request=<...>` query parameter.
String signedLocalRpLoginRequestToUrlParam(SignedLocalRpLoginRequest signed) =>
    encodeUrlParam(Codec.encodeSignedLocalRpLoginRequest(signed));

SignedLocalRpLoginRequest signedLocalRpLoginRequestFromUrlParam(String param) =>
    Codec.decodeSignedLocalRpLoginRequest(decodeUrlParam(param));

/// Encode for the callback redirect's `&encrypted_token=<...>` query
/// parameter.
String localRpEncryptedCallbackToUrlParam(LocalRpEncryptedCallback callback) =>
    encodeUrlParam(Codec.encodeLocalRpEncryptedCallback(callback));

LocalRpEncryptedCallback localRpEncryptedCallbackFromUrlParam(String param) =>
    Codec.decodeLocalRpEncryptedCallback(decodeUrlParam(param));
